pkg/apis/extensions/types.go

```diff
@@ -82,29 +82,48 @@ type SubresourceReference struct {
 	Subresource string `json:"subresource,omitempty"`
 }
 
-// ResourceConsumption is an object for specifying average resource consumption of a particular resource.
-type ResourceConsumption struct {
-	Resource api.ResourceName  `json:"resource,omitempty"`
-	Quantity resource.Quantity `json:"quantity,omitempty"`
+// MetricName is an identifier of a Heapster metric.
+type MetricName string
+
+const (
+	// Heapster metric representing CPU usage of a pod (in milli-cores).
+	MetricCPUUsage MetricName = "kubernetes.io:cpu-usage"
+	// Heapster metric representing Memory usage of a pod (in bytes).
+	MetricMemoryUsage MetricName = "kubernetes.io:memory-usage"
+)
+
+// MetricUtilization is an object for specifying the arthmetic mean utilization of a particular resource.
+type MetricUtilization struct {
+	// Metric specifies either the identifier of the target metric when present in the spec, or the identifier of the observed metric when present in the status.
+	Metric MetricName `json:"metric"`
+	// Utilization specifies either the target arthmetic mean utilization of the metric when present in the spec, or the observed arthmetic mean utilization when present in the status.
+	Utilization resource.Quantity `json:"utilization"`
 }
 
 // HorizontalPodAutoscalerSpec is the specification of a horizontal pod autoscaler.
 type HorizontalPodAutoscalerSpec struct {
 	// ScaleRef is a reference to Scale subresource. HorizontalPodAutoscaler will learn the current resource consumption from its status,
 	// and will set the desired number of pods by modyfying its spec.
-	ScaleRef *SubresourceReference `json:"scaleRef"`
+	ScaleRef SubresourceReference `json:"scaleRef"`
 	// MinReplicas is the lower limit for the number of pods that can be set by the autoscaler.
 	MinReplicas int `json:"minReplicas"`
 	// MaxReplicas is the upper limit for the number of pods that can be set by the autoscaler. It cannot be smaller than MinReplicas.
 	MaxReplicas int `json:"maxReplicas"`
-	// Target is the target average consumption of the given resource that the autoscaler will try to maintain by adjusting the desired number of pods.
-	// Currently two types of resources are supported: "cpu" and "memory".
-	Target ResourceConsumption `json:"target"`
+	// TargetMetricUtilizations are the target arthmetic mean utilizations of the given metrics that the autoscaler will try to maintain by adjusting the desired number of pods.
+	// Currently two types of metrics are supported: "kubernetes.io:cpu-usage" and "kubernetes.io:memory-usage".
```

Comment:
Is memory actually supported?

Please omit `kubernetes.io:`.

Please define `cpu-usage` more precisely (e.g., average vs. median, time window), or add a caveat that its precise meaning is likely to change.